### ISSUE TITLE
Introduce `reportUnnecessaryTypeIs` rule

### ISSUE THREAD

**Issue by kkom** [microsoft/pyright#7990]
**Is your feature request related to a problem? Please describe.**

Consider the following code:

```python
class Undefined:
   """Sentinel value analogous to TypeScript's `undefined`."""

UNDEFINED = Undefined()
"""Canonical singleton value of the Undefined type."""
```

The `Undefined` class is something we already have in our codebase. We currently compare values against it using the `if isinstance(value, Undefined):` syntax which works, but is a bit verbose and complex.

When [PEP 742](https://peps.python.org/pep-0742/) was accepted, we wanted to introduce a concise helper for it:

```python
from typing_extensions import TypeIs

def is_undefined(x: object) -> TypeIs[Undefined]:
    """Canonical and concise way to check if a value is Undefined."""

    return isinstance(x, Undefined)
```

However, the fact that `TypeIs` doesn't have a rule analogous to `isinstance`'s `reportUnnecessaryIsInstance` stops us from doing it (because of the risk of missing dead code).

**Describe the solution you’d like**

Would it be possible to introduce a `reportUnnecessaryTypeIs` rule that would fire whenever the type of the value passed to a function using of the `TypeIs` special form is already equal to the narrowed type?

**Comment by erictraut** [comment]
Such a feature is theoretically possible, but I'm skeptical that users other than you would find it useful. 

I'm going to reject the feature request for now, but I'm willing to reconsider in the future if there is sufficient signal from other pyright users (in the form of upvotes on this issue).

**Comment by kkom** [comment]
Thank you so much for the quick response Eric!

I agree that the example I gave wasn't very compelling. In fact, in this case we're actually happy continuing to inline the `isinstance(..., Undefined)` checks - proving your point. However, I'd like to offer a slightly broader perspective which hopefully may make you more willing to reconsider it (in addition to any upvotes).

Of course, I understand that the decision is yours and I'll wait for further feedback from the community.

#### Less trivial uses of `TypeIs`

We've got other, less trivial, uses for the `TypeIs` / `TypeForm` special forms. For example, the `NonEmptyString` type which is implemented approximately like this:

```python
NonEmptyString = NewType("NonEmptyString", str)

def is_non_empty_string(value: str | None) -> TypeIs[NonEmptyString]:
    if value is None:
        return False

    return len(value) > 0
```

We're using it to avoid continuously repeating a common data validation check of "is this string longer than 0" – which adds both runtime overhead, but more importantly - requires us to branch the logic of our application code.

In this case, a rule like `reportUnnecessaryTypeIs` would be very helpful when refactoring our code and moving data validation to happen sooner – ensuring that unnecessary branching is eliminated along the way.

#### Analogy to `reportUnnecessaryIsInstance`, `reportUnnecessaryCast` & `reportUnnecessaryComparison`

Even more broadly, I see `TypeIs` as a way to extend the already established flow-sensitive typing functionality of Pyright beyond types literals / enums / `None` / etc.

Since rules like `reportUnnecessaryIsInstance`, `reportUnnecessaryCast` & `reportUnnecessaryComparison` already exist, it feels to me like `reportUnnecessaryTypeIs` would fit coherently with them, while providing useful support for broader adoption of [PEP 742](https://peps.python.org/pep-0742/).

**Comment by erictraut** [comment]
The existing `reportUnnecessaryXXX` rules were implemented because these conditions were proven to represent common sources of bugs. I would be surprised if intentional (but unnecessary) calls to type guard functions currently represent a common source of bugs. As I said, I'm willing to revisit this in the future based on signal from pyright users.

If we decide to implement this at some point, I would first think about adding the new check under the existing `reportUnnecessaryIsInstance` diagnostic rule. It already covers both `isinstance` and `issubclass` calls, so it could be extended to `TypeIs` calls. Adding new diagnostic rules is something we do very reluctantly and cautiously because they increase complexity and are difficult to change or remove. Extending existing rules has a much lower threshold.
